cou/steps/nova_cloud_controller.py

```diff
@@ -0,0 +1,89 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Functions for prereq steps relating to nova."""
+import logging
+
+from cou.exceptions import ApplicationNotFound, COUException, UnitNotFound
+from cou.utils.juju_utils import Model
+
+logger = logging.getLogger(__name__)
+
+
+async def archive(model: Model, *, batch_size: int) -> None:
+    """Archive data on a nova-cloud-controller unit.
+
+    The archive-data action only runs a single batch,
+    so we run it in a loop until completion.
+    See also:
+    https://charmhub.io/nova-cloud-controller/actions#archive-data
+    https://docs.openstack.org/project-deploy-guide/charm-deployment-guide/wallaby/upgrade-openstack.html#archive-old-database-data
+    :param model: juju model to work with
+    :type model: Model
+    :param batch_size: batch-size to pass to the archive-data action
+        (default is 1000; decrease if performance issues seen)
+    :type batch_size: int
+    :raises COUException: if action returned unexpected output
+    """  # noqa: E501 line too long
+    unit_name: str = await _get_nova_cloud_controller_unit_name(model)
+    # The archive-data action only archives a single batch,
+    # so we must run it in a loop until everything is archived.
+    while True:
+        logger.debug("Running action 'archive-data' on %s", unit_name)
+        # https://charmhub.io/nova-cloud-controller/actions#archive-data
+        action = await model.run_action(
+            unit_name=unit_name,
+            action_name="archive-data",
+            raise_on_failure=True,
+            action_params={"batch-size": batch_size},
+        )
+        logger.info("action output: %s", action.data)
+        output = action.data["results"].get("archive-deleted-rows")
+        if output is None:
+            raise COUException(
+                "Expected to find output in action results.'archive-deleted-rows', "
+                "but it was not present."
+            )
+        # The command will contain this string if there is nothing left to archive.
+        # This means we don't need to run the command any more.
+        if "Nothing was archived" in output:
+            logger.debug("Archiving complete.")
+            break
+        logger.debug("Potentially more data to archive...")
+
+
+async def _get_nova_cloud_controller_unit_name(model: Model) -> str:
+    """Get nova-cloud-controller application's first unit's name.
+
+    Assumes only a single nova-cloud-controller application is deployed.
+
+    :param model: juju model to work with
+    :type model: Model
+    :return: unit name
+    :rtype: str
+    :raises UnitNotFound: When cannot find a valid unit for 'nova-cloud-controller'
+    :raises ApplicationNotFound: When cannot find a 'nova-cloud-controller' application
+    """
+    status = await model.get_status()
```

Comment:
Actually, why we need to use status here? This operation is time-consuming, so I think we can simply iterate over apps in model.

```python
fro app in model.appications.values():
    if app.charm_name == "nova-cloud-controller":
        if app.units:
            return app.units[0].name
...
```